### PR TITLE
Add support for AzureOpenAITextModelCategoricalScore

### PR DESCRIPTION
**Pull request by juliagongms**
Adds support for an evaluator of the `captioning` task that uses single-turn text-only calls to an Azure OpenAI model for evaluation and returns standard categorical metrics (precision, recall, F1, accuracy) based on the scores returned, along with an average model-assigned score.

### CODE REVIEW

File: visionmetrics/caption/azure_openai_model_eval_base.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+                        an exact match for either a prediction or target to be considered negative. TODO: Implement fuzzy negative matching. Scores are always computed so they
+                        can be used for average_score computation.
+
+        Other parameters follow the standards of irisml.tasks.create_azure_openai_chat_model.OpenAITextChatModel.
+    """
+    def __init__(self, endpoint: str, deployment_name: str, system_message: str, prompt_template: str,
+                 temperature=0.0, max_tokens=50, requests_interval=30, num_responses=1, positive_threshold=0.5, negative_value=''):
+        super().__init__()
+        if PREDICTION_PLACEHOLDER not in prompt_template or TARGET_PLACEHOLDER not in prompt_template:
+            raise ValueError("Both the predicted placeholder {PREDICTION_PLACEHOLDER} and target placeholder {TARGET_PLACEHOLDER} must be present in prompt_template.")
```
Comment (nisyad-ms) [inline visionmetrics/caption/azure_openai_model_eval_base.py:46]: If both placeholders must be present, then maybe you meant to use AND in line #45?

File: visionmetrics/caption/azure_openai_model_eval_base.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    def _get_numeric_score(self, raw_score: str):
+        """
+        Parses the numeric score value from the string `raw_score`; if not parseable, returns 0.
+        """
+        score = 0.
+        try:
+            score = float(raw_score)
+        except ValueError:
+            self.score_parse_failures += 1
```
Comment (cy-bai) [inline visionmetrics/caption/azure_openai_model_eval_base.py:76]: can we logger.DEBUG the raw_score here?

File: visionmetrics/caption/azure_openai_model_eval.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class AzureOpenAITextModelCategoricalScore(AzureOpenAITextModelCategoricalEvalBase):
+    """
+    Prompt-tuned evaluator that uses single-turn text-only calls to an Azure OpenAI model for evaluation and returns
+    standard categorical metrics (precision, recall, F1, accuracy) based on the scores returned.
+    """
+
+    def __init__(self):
+        self.system_message = """You are an expert evaluator of AI models. Follow the steps below:
```
Comment (cy-bai) [inline visionmetrics/caption/azure_openai_model_eval.py:11]: Do we expect tuning the prompts in experiments? If so, maybe we can make this system message and prompt as default values, and add an argument to customize it?

**Review by cy-bai — COMMENTED** [review]
Added a few general questions

File: visionmetrics/caption/azure_openai_model_eval_base.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        self.predictions += predictions
+        self.targets += targets
+        
+        for prediction, target in zip(predictions, targets):
+            prompt = self.prompt_template.replace(PREDICTION_PLACEHOLDER, prediction)
+            final_target = OR_SEPARATOR.join(target)
+            prompt = prompt.replace(TARGET_PLACEHOLDER, final_target)
+            
+            try:
+                result = ["1.0"] if prediction == target else self.model.forward(([prompt], [[]]))
```
Comment (nisyad-ms) [inline visionmetrics/caption/azure_openai_model_eval_base.py:99]: nit: self.model([prompt], [[]]) should suffice if it is inheriting from nn.Module?

File: visionmetrics/caption/azure_openai_model_eval_base.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
+import torch
+from collections import Counter
+from enum import Enum
+from irisml.tasks.create_azure_openai_chat_model import OpenAITextChatModel
```
Comment (nisyad-ms) [inline visionmetrics/caption/azure_openai_model_eval_base.py:4]: There is also `irisml.tasks.create_azure_openai_completions_model` but looking at the code it look like both call the /completions API?

File: visionmetrics/caption/azure_openai_model_eval.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+Prediction: The bar chart visualizes the cost in dollars across different healthcare categories, along with the number of patients covered, average premiums, and number of insurers. The healthcare categories include Private Health Insurance with 9,500 patients, Outpatient Care with 8,500 patients, Inpatient Care with 6,000 patients, Mental Health Care with 4,000 patients, Dental Care with 3,000 patients, Vision Care with 1,000 patients, and Prescription Drugs with 5,000 patients. Costs are represented by blue bars, patients covered by a blue line, average premiums by pink markers connected with lines, and insurers indicated on the secondary y-axis.
+Score: 0.6
+
+Ground truth: 3 dogs
+Prediction: 4 dogs
+Score: 0.2
```
Comment (nisyad-ms) [inline visionmetrics/caption/azure_openai_model_eval.py:22]: Why is score=0.2?

File: visionmetrics/caption/azure_openai_model_eval_base.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        for prediction, target in zip(predictions, targets):
+            prompt = self.prompt_template.replace(PREDICTION_PLACEHOLDER, prediction)
+            final_target = OR_SEPARATOR.join(target)
+            prompt = prompt.replace(TARGET_PLACEHOLDER, final_target)
+            
+            try:
+                result = ["1.0"] if prediction == target else self.model.forward(([prompt], [[]]))
+                result = result[0] if result else ''
+            except:
+                result = '0.0'
```
Comment (nisyad-ms) [inline visionmetrics/caption/azure_openai_model_eval_base.py:102]: Any particular reason why result here = '0.0' and not '' (empty str)
Also, line 99 may never fail since the chat completion model in irisml already catches exception and returns empty str.

File: visionmetrics/caption/azure_openai_model_eval.py (hunk 3 above)
Comment (nisyad-ms) [reply visionmetrics/caption/azure_openai_model_eval.py:11]: Maybe define this SYSTEM PROMPT outside the class as a constant - if the user doesn't provide their system prompt, then default gets used as Chongyang suggested.

File: visionmetrics/caption/azure_openai_model_eval_base.py (hunk 7 above)
Comment (juliagongms) [reply visionmetrics/caption/azure_openai_model_eval_base.py]: Good catch, thanks! Simplified this logic.

File: visionmetrics/caption/azure_openai_model_eval.py (hunk 3 above)
Comment (juliagongms) [reply visionmetrics/caption/azure_openai_model_eval.py]: GMTA, was thinking the same thing today. Updated!

File: visionmetrics/caption/azure_openai_model_eval_base.py (hunk 5 above)
Comment (juliagongms) [reply visionmetrics/caption/azure_openai_model_eval_base.py:4]: The completions and chat functionalities are different, and our methods use the chat function (the completions function is more outdated and not instruction-tuned, and the newer models like GPT-4o don't support it). The chat completions API is /chat/completions, while the /completions API is different (https://github.com/microsoft/irisml-tasks-azure-openai/blob/main/irisml/tasks/create_azure_openai_completion_model.py#L157 vs https://github.com/microsoft/irisml-tasks-azure-openai/blob/main/irisml/tasks/create_azure_openai_chat_model.py#L81).

File: visionmetrics/caption/azure_openai_model_eval.py (hunk 6 above)
Comment (juliagongms) [reply visionmetrics/caption/azure_openai_model_eval.py:28]: This is an initial crude prompt that worked reasonably well, but more prompt fine-tuning will happen in the coming days. The in-context examples are loose guidelines. My reasoning for this example was to balance the ambiguity that occurs in the vision space - for instance, 3 dogs vs. 4 dogs in a license plate reading task vs a detection/counting task may merit varying levels of partial credit. I hope to create some more detailed guidelines in the next iteration of the default prompt to capture a more systematic way of reasoning about the scores.

Comment (nisyad-ms) [reply visionmetrics/caption/azure_openai_model_eval.py:28]: If needed in the future, we can store different default prompts in a system_prompt.json file

File: visionmetrics/caption/azure_openai_model_eval_base.py (hunk 5 above)
Comment (nisyad-ms) [reply visionmetrics/caption/azure_openai_model_eval_base.py:4]: Aha! Those two are indeed different :).

**Review by nisyad-ms — APPROVED** [review]
